src/modules/alertmanager/CriticalAlertsService.ts

```diff
@@ -0,0 +1,171 @@
+            result[operator.nos_name] = {all: proposeStats.all, missed: proposeStats.missed};
+          }
+        }
+        return result;
+      };
+      const sendRule = (ruleResult: AlertRuleResult): boolean => {
+        const defaultInterval = 6 * 60 * 60 * 1000; // 6h
+        const ifIncreasedInterval = 60 * 60 * 1000; // 1h
+        sendTimestamp = Date.now();
+        if (Object.values(ruleResult).length > 0) {
+          const prevSendTimestamp = sentAlerts[alertname]?.timestamp ?? 0;
+          if (sendTimestamp - prevSendTimestamp > defaultInterval) return true;
+          for (const [operator, operatorResult] of Object.entries(ruleResult)) {
+            const prevAll = sentAlerts[alertname]?.ruleResult[operator]?.all ?? 0;
+            const prevMissed = sentAlerts[alertname]?.ruleResult[operator]?.missed ?? 0;
+            // if math relation of missed to all increased
+            if (
+              (operatorResult.missed / operatorResult.all) > (prevMissed / prevAll) &&
+              sendTimestamp - prevSendTimestamp > ifIncreasedInterval
+            ) return true;
+          }
+        }
+        return false;
+      }
+      const alertBody = (ruleResult: AlertRuleResult): AlertRequestBody => {
+        return {
+          startsAt: new Date(sendTimestamp).toISOString(),
+          endsAt: new Date(new Date(sendTimestamp).setMinutes(new Date(sendTimestamp).getMinutes() + 1)).toISOString(),
+          labels: {alertname, severity: 'critical'},
+          annotations: {
+            summary: `${Object.values(ruleResult).length} Node Operators with CRITICAL count of missed proposes in the last 12 hours`,
+            description: join(Object.entries(ruleResult).map(([o, r]) => `${o}: ${r.negDelta} of ${r.ongoing} proposes`), '\n'),
+          }
+        }
+      };
+      const toSend = async (): Promise<PreparedToSendAlert | undefined> => {
+        const ruleResult = await alertRule();
+        if (sendRule(ruleResult)) return {timestamp: sendTimestamp, body: alertBody(ruleResult), ruleResult};
+      }
+      return toSend();
+    }
+  ]
+
+  public constructor(
+    @inject(ILogger) protected logger: ILogger,
+    @inject(Environment) protected env: Environment,
+    @inject(ClickhouseStorage) protected storage: ClickhouseStorage,
+    @inject(DataProcessor) protected dataProcessor: DataProcessor,
+    @inject(Environment) protected environment: Environment
+  ) {
+    this.baseUrl = this.env.ALERTMANAGER_URL ?? '';
+  }
+
+  public async sendCriticalAlerts(slot: bigint) {
+    if (!this.baseUrl) {
+      this.logger.warn(`Env var 'ALERTMANAGER_URL' is not set. Unable to send critical alerts`)
+      return;
+    }
+    this.logger.info('Send critical alerts if exist')
+    try {
+      for (const alert of this.alerts) {
+        const toSend = await alert(slot);
+        if (toSend) await this.push(toSend.body).then(() => sentAlerts[toSend.body.labels.alertname] = toSend)
```

Comment:
Why we need `sentAlerts`?

Reply:
It's just last sent alerts registry. To manage sending intervals